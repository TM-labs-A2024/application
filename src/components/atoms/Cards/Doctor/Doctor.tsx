import { Link } from '@chakra-ui/next-js'
import { Text } from '@chakra-ui/react'
import { specialties } from '@constants/index'
import { Doctor as DoctorType } from '@src/types'
import NextLink from 'next/link'
import React from 'react'

import Chevron from '../../../../../public/static/icons/chevron.svg'

export default function Doctor({ doctor }: { doctor: DoctorType }) {
  // --- Data and handlers -----------------------------------------------------
  const doctorSpecialtiesList = doctor.specialties?.map(
    (specialty, idx) =>
      `${specialties.find((el) => el.id === specialty)?.name}${
        doctor.specialties.length > 0 && idx !== doctor.specialties.length - 1 ? ',' : ''
      }${idx === doctor.specialties.length - 1 ? '.' : ' '}`
  )
  // --- END: Data and handlers ------------------------------------------------

  return (
    <Link
      as={NextLink}
      href={`/medico/${doctor.id}/${doctor.requestId}`}
      className="w-full"
      data-testid="doctor-card"
    >
      <div className="flex flex-row items-center justify-between">
        <div className="flex flex-col items-start gap-1 overflow-hidden">
          <Text>
            {doctor.firstname} {doctor.lastname}
          </Text>
          <div className="flex flex-row flex-wrap items-center gap-2">
            <Text className="text-nowrap">CI: {doctor.govId}</Text>
            <Text className="text-wrap">{doctorSpecialtiesList}</Text>
          </div>
        </div>
        <Chevron />
      </div>
    </Link>
  )
}
