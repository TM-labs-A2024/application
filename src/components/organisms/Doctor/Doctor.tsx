import { ArrowBackIcon } from '@chakra-ui/icons'
import {
  IconButton,
  Text,
  Stack,
  Heading,
  Divider,
  Button,
  Modal,
  ModalOverlay,
  ModalContent,
  ModalFooter,
  ModalBody,
  Spinner
} from '@chakra-ui/react'
import { specialties } from '@constants/index'
import { Doctor as DoctorType } from '@src/types'
import { isIOS } from '@utils/index'
import { format } from 'date-fns'
import { es } from 'date-fns/locale/es'
import { useRouter } from 'next/navigation'
import React, { ReactElement } from 'react'

function ConfirmationModal({
  isOpen,
  onClose,
  onSubmit,
  method,
  name
}: {
  isOpen: boolean
  onClose: () => void
  onSubmit: () => void
  method: string
  name: string
}) {
  return (
    <Modal isOpen={isOpen} onClose={onClose}>
      <ModalOverlay />
      <ModalContent>
        <ModalBody>
          ¿Seguro que desea {method.toLowerCase()} el acceso a {name}?
        </ModalBody>

        <ModalFooter>
          <Button mr={3} onClick={onClose}>
            Cancelar
          </Button>
          <Button colorScheme="blackAlpha" onClick={onSubmit}>
            {method} acceso
          </Button>
        </ModalFooter>
      </ModalContent>
    </Modal>
  )
}

export default function Doctor({
  context: {
    doctor,
    isDenialOpen,
    onDenialOpen,
    onDenialClose,
    isApprovalOpen,
    onApprovalOpen,
    onApprovalClose,
    isRemovalOpen,
    onRemovalOpen,
    onRemovalClose,
    onApproval,
    onDenial,
    onRemoval,
    isLoading
  }
}: {
  context: {
    doctor: DoctorType
    isDenialOpen: boolean
    onDenialOpen: () => void
    onDenialClose: () => void
    isApprovalOpen: boolean
    onApprovalOpen: () => void
    onApprovalClose: () => void
    isRemovalOpen: boolean
    onRemovalOpen: () => void
    onRemovalClose: () => void
    onApproval: () => void
    onDenial: () => void
    onRemoval: () => void
    isLoading: boolean
  }
}): ReactElement {
  // --- Hooks -----------------------------------------------------------------
  const router = useRouter()
  // --- END: Hooks ------------------------------------------------------------

  // --- Local state -----------------------------------------------------------
  // --- END: Local state ------------------------------------------------------

  // --- Refs ------------------------------------------------------------------
  // --- END: Refs -------------------------------------------------------------

  // --- Redux -----------------------------------------------------------------
  // --- END: Redux ------------------------------------------------------------

  // --- Side effects ----------------------------------------------------------
  // --- END: Side effects -----------------------------------------------------

  // --- Data and handlers -----------------------------------------------------
  const doctorSpecialtiesList = doctor.specialties?.map(
    (specialty, idx) =>
      `${specialties.find((el: { id: string; name: string }) => el.id === specialty)?.name}${
        doctor.specialties.length > 0 && idx !== doctor.specialties.length - 1 ? ',' : ''
      }${idx === doctor.specialties.length - 1 ? '.' : ' '}`
  )
  // --- END: Data and handlers ------------------------------------------------

  return (
    <div
      className={`flex h-screen w-screen flex-col p-8 lg:px-96 ${isIOS() ? 'pt-20' : 'pt-8'}`}
      data-testid="doctor"
    >
      <div className="mb-8 flex flex-row items-center justify-start">
        <IconButton
          size="xl"
          className="mr-4"
          aria-label="back"
          variant="link"
          icon={<ArrowBackIcon />}
          onClick={() => {
            router.push('/medicos')
          }}
        />
        <Text className="font-medium">Perfil del médico</Text>
      </div>
      <div className="flex h-full flex-col justify-between">
        <div>
          <Stack spacing={1} mb={6}>
            <Heading as="h3" size="md" noOfLines={1}>
              {doctor.firstname} {doctor.lastname}
            </Heading>
            <Text>CI: {doctor.govId}</Text>
          </Stack>
          <Divider orientation="horizontal" />
          <Stack mb={6} mt={6}>
            <h4 className="text-sm text-gray-600">Especialidad</h4>
            <Text className="text-wrap">{doctorSpecialtiesList}</Text>
          </Stack>
          <Stack mb={6}>
            <h4 className="text-sm text-gray-600">Fecha de nacimiento</h4>
            <Text className="font-medium">
              {format(new Date(doctor.birthdate), "dd 'de' MMMM, yyyy", {
                locale: es
              })}
            </Text>
          </Stack>
          <Stack mb={6}>
            <h4 className="text-sm text-gray-600">Correo electrónico</h4>
            <Text className="font-medium">{doctor.email}</Text>
          </Stack>
          <Stack mb={6}>
            <h4 className="text-sm text-gray-600">Teléfono</h4>
            <Text className="font-medium">{doctor.phoneNumber}</Text>
          </Stack>
        </div>
        {isLoading && <Spinner />}
        {!doctor?.patientPending && !isLoading && (
          <Button onClick={onRemovalOpen}>Revocar acceso</Button>
        )}
        {doctor?.patientPending && !isLoading && (
          <div className="flex w-full flex-row">
            <Button className="mr-2 flex-grow" variant="outline" onClick={onDenialOpen}>
              Rechazar
            </Button>
            <Button className="ml-2 flex-grow" onClick={onApprovalOpen}>
              Aceptar
            </Button>
          </div>
        )}
      </div>
      <ConfirmationModal
        isOpen={isDenialOpen}
        onClose={onDenialClose}
        onSubmit={onDenial}
        method="Denegar"
        name={`${doctor.firstname} ${doctor.lastname}`}
      />
      <ConfirmationModal
        isOpen={isApprovalOpen}
        onClose={onApprovalClose}
        onSubmit={onApproval}
        method="Conceder"
        name={`${doctor.firstname} ${doctor.lastname}`}
      />
      <ConfirmationModal
        isOpen={isRemovalOpen}
        onClose={onRemovalClose}
        onSubmit={onRemoval}
        method="Revocar"
        name={`${doctor.firstname} ${doctor.lastname}`}
      />
    </div>
  )
}
