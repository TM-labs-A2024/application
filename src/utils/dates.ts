export const formatDate = (date: string) => {
  const chunks = date.split('-')

  const newDate = chunks[1] + '-' + chunks[2] + '-' + chunks[0]

  return newDate
}
