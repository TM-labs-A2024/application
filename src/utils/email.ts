import emailjs from '@emailjs/browser'

export function sendEmail(data: {
  from_name: string
  to_name: string
  code: string | number
  to_email: string
}) {
  const templateParams = {
    ...data
  }

  const serviceKey = process.env.NEXT_PUBLIC_EMAIL_SERVICE_KEY ?? ''
  const templateKey = process.env.NEXT_PUBLIC_EMAIL_TEMPLATE_KEY ?? ''
  const apiKey = process.env.NEXT_PUBLIC_EMAIL_API_KEY ?? ''

  emailjs.send(serviceKey, templateKey, templateParams, apiKey).catch(() => {
    alert('Algo sali√≥ mal, por favor reintenta el registro.')
  })
}
