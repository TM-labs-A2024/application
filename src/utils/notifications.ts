import { GENERIC_ERROR } from '@constants/index'
import { isMobile } from '@utils/index'
import { Store } from 'react-notifications-component'

export const setupErrorNotification = () => {
  Store.addNotification(
    GENERIC_ERROR('Algo salió mal, por favor inténtelo de nuevo.', isMobile(window))
  )
}
