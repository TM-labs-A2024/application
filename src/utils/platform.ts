import { Capacitor } from '@capacitor/core'

export const isWeb = () => Capacitor.getPlatform() === 'web'

export const isIOS = () => Capacitor.getPlatform() === 'ios'

export const isAndroid = () => Capacitor.getPlatform() === 'android'
