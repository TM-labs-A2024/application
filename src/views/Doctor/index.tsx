import Doctor from '@components/organisms/Doctor'
import { Doctor as DoctorType } from '@src/types'
import React from 'react'

export default function DoctorView({
  context
}: {
  context: {
    doctor: DoctorType
    isDenialOpen: boolean
    onDenialOpen: () => void
    onDenialClose: () => void
    isApprovalOpen: boolean
    onApprovalOpen: () => void
    onApprovalClose: () => void
    isRemovalOpen: boolean
    onRemovalOpen: () => void
    onRemovalClose: () => void
    onApproval: () => void
    onDenial: () => void
    onRemoval: () => void
    isLoading: boolean
  }
}) {
  return <Doctor context={context} />
}
