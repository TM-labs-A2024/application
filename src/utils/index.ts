import { formatDate } from './dates'
import { sendEmail, generateVerificationCode, setupEmailSending } from './email'
import { setupErrorNotification } from './notifications'
import { isWeb, isIOS, isAndroid, isMobile } from './platform'

export {
  isWeb,
  isIOS,
  isAndroid,
  isMobile,
  sendEmail,
  generateVerificationCode,
  setupEmailSending,
  setupErrorNotification,
  formatDate
}
