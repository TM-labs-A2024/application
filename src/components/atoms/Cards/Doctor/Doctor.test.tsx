import { doctors } from '@constants/index'
import { render, screen } from '@test/utils/index'

import '@testing-library/jest-dom'
import Doctor from './Doctor'

describe('Atoms > Cards > Doctor test', () => {
  test('Loads and displays doctor name', async () => {
    // ARRANGE
    render(<Doctor doctor={doctors[0]} />)

    // ACT
    await screen.findByText('José Pérez')

    // ASSERT
    expect(screen.getByText('José Pérez')).toBeInTheDocument()
  })

  test('Match the snapshot', async () => {
    render(<Doctor doctor={doctors[0]} />)

    expect(await screen.findByTestId('doctor-card')).toMatchSnapshot()
  })
})
