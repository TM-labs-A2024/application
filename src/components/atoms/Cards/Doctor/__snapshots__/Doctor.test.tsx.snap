// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atoms > Cards > Doctor test Matches the snapshot 1`] = `
<a
  class="chakra-link w-full css-spn4bz"
  data-testid="doctor-card"
  href="/medico/1000000/1000000"
>
  <div
    class="flex flex-row items-center justify-between"
  >
    <div
      class="flex flex-col items-start gap-1 overflow-hidden"
    >
      <p
        class="chakra-text css-0"
      >
        José
         
        Pérez
      </p>
      <div
        class="flex flex-row flex-wrap items-center gap-2"
      >
        <p
          class="chakra-text text-nowrap css-0"
        >
          CI: 
          1000000
        </p>
        <p
          class="chakra-text text-wrap css-0"
        >
          Alergología, 
          Cardiología.
        </p>
      </div>
    </div>
    <div />
  </div>
</a>
`
